### PR TITLE
Fix job control if ksh is started as session leader

### PR DESCRIPTION
**Pull request by siteshwar**
Ksh fails to initalize job control if it is started as login shell. A
login shell is started as a session leader. Ksh tries to verify if job
control is set up properly by calling setpgid() on it's own process.
According to POSIX, calling setpgid() on a session leader should fail
with EPERM. This was causing ksh to skip initialzing job control.

We should be able to initalize job control by safely assuming that ksh
is started in a valid process group.

Resolves: #437

### CODE REVIEW

File: src/cmd/ksh93/sh/jobs.c [hunk 1]
```diff
@@ -598,40 +597,17 @@ void job_init(Shell_t *shp, int lflag) {
 #endif /* NTTYDISC */
 
     job.mypgid = getpgrp();
-    // Some systems have job control, but not initialized.
-    if (job.mypgid <= 0) {
```
Comment (siteshwar) [inline src/cmd/ksh93/sh/jobs.c]: I am not sure about other systems, but on Linux `getpgrp()` should return 0 only if ksh is started as first process i.e. pid 1. Given that is it safe to remove this code @msekletar ?

Comment (siteshwar) [reply src/cmd/ksh93/sh/jobs.c]: Pid 1 has process group 1 on Linux, so my previous comment was incorrect.

File: src/cmd/ksh93/sh/jobs.c [hunk 2]
```diff
@@ -598,40 +597,17 @@ void job_init(Shell_t *shp, int lflag) {
...
-        if ((fd = sh_open(ttynam, O_RDWR)) < 0) return;
-        if (fd != JOBTTY) sh_iorenumber(shp, fd, JOBTTY);
-        job.mypgid = shp->gd->pid;
-#ifdef SIGTSTP
-        tcsetpgrp(JOBTTY, shp->gd->pid);
-        setpgid(0, shp->gd->pid);
-#endif  // SIGTSTP
-    }
 #ifdef SIGTSTP
-    possible = setpgid(0, job.mypgid) >= 0;
```
Comment (siteshwar) [inline src/cmd/ksh93/sh/jobs.c]: This call to `setpgid()` succeeds in last stable version of ksh. I am still checking what's causing it to fail now.

Comment (siteshwar) [reply src/cmd/ksh93/sh/jobs.c]: `setpgid()` is actually failing on last stable version of ksh. This bug was introduced by reformatting changes in 00d0a53b8.

This line:
```
    if(possible = (setpgid(0,job.mypgid)>=0) || errno==EPERM)     
```
was changed to:

```
   possible = setpgid(0, job.mypgid) >= 0;
    if (possible || errno == EPERM) {
```
So `possible` is not set to 1 if `setpgid()` fails due to `EPERM`. It should have been changed to:

```
   possible = setpgid(0, job.mypgid) >= 0  || errno == EPERM;
    if (possible) {
```

Comment (krader1961) [reply src/cmd/ksh93/sh/jobs.c]: Good catch and this is a good example for why the existing style that does assignment inside `if()` conditions is a really bad idea. Also, the `>= 0` test is itself confusing since `setpgid()` by definition returns zero on success, -1 on failure, and should never return a positive value. It would be clearer to write it as

```c
possible = setpgid(0, job.mypgid) != -1 || errno == EPERM;
```
